internal/gatewayapi/listeners.go

```diff
@@ -0,0 +1,136 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayapi
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayapi_v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+type ValidateListenersResult struct {
+	InsecurePort int
+	SecurePort   int
+
+	InvalidListenerConditions map[gatewayapi_v1alpha2.SectionName]metav1.Condition
+}
+
+// ValidateListeners validates protocols, ports and hostnames on a set of listeners.
+// It ensures that:
+//	- all protocols are supported
+//	- each listener group (grouped by protocol, with HTTPS & TLS going together) uses a single port
+//  - hostnames within each listener group are unique
+// It returns the insecure & secure ports to use, as well as conditions for all invalid listeners.
+// If a listener is not in the "InvalidListenerConditions" map, it is assumed to be valid according
+// to the above rules.
+func ValidateListeners(listeners []gatewayapi_v1alpha2.Listener) ValidateListenersResult {
+	result := ValidateListenersResult{
+		InvalidListenerConditions: map[gatewayapi_v1alpha2.SectionName]metav1.Condition{},
+	}
+
+	// All listeners with a protocol of "HTTP" must use the same port number
+	// Heuristic: the first port number encountered is allowed, any other listeners with a different port number are marked "Detached" with "PortUnavailable"
```

Comment:
this may be something to call out in a doc somewhere, i dont think this specific grouping/heuristic is in the gateway docs themselves (looking at the `Listeners` field documentation here https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.Gateway)

Reply:
Yeah, this is really a Contour-specific thing since we only support 2 ports total. Agree we should call this out somewhere - need to figure out where to put docs for all this, but let me add this to a new issue.